clients/githubrepo/client.go

```diff
@@ -15,48 +15,89 @@
 package githubrepo
 
 import (
+	"archive/tar"
+	"compress/gzip"
 	"context"
 	"fmt"
 	"io"
 	"io/ioutil"
+	"log"
 	"net/http"
 	"os"
+	"path/filepath"
 	"strings"
 
 	"github.com/google/go-github/v32/github"
+	"github.com/pkg/errors"
 
 	"github.com/ossf/scorecard/clients"
 )
 
-const repoFilename = "githubrepo*.tar.gz"
+const (
+	repoDir      = "repo*"
+	repoFilename = "githubrepo*.tar.gz"
+)
+
+var errZipSlip = errors.New("ZipSlip path detected")
+
+func extractAndValidateArchivePath(path, dest string) (string, error) {
+	const splitLength = 2
+	names := strings.SplitN(path, "/", splitLength)
+	if len(names) < splitLength {
+		log.Printf("Unable to split path: %s", path)
+		return "", nil
+	}
+	// Check for ZipSlip: https://snyk.io/research/zip-slip-vulnerability
+	cleanpath := filepath.Join(dest, names[1])
+	if !strings.HasPrefix(cleanpath, dest) {
```

Comment:
does `dest` end with "/" ? 

Otherwise if `dest` is `/tmp`, we can write to `/tmpblah` as well? 

probably better to clean `dest` too and add a "/".

Also, how about returning the joined value here (if it's valid), rather than discarding the joined value (`dest` is not really necessary at this point), rather than having the caller join this again?

Reply:
> does `dest` end with "/" ?
> 
> Otherwise if `dest` is `/tmp`, we can write to `/tmpblah` as well?
> 
> probably better to clean `dest` too and add a "/".
> 

Nice catch! Done.

> Also, how about returning the joined value here (if it's valid), rather than discarding the joined value (`dest` is not really necessary at this point), rather than having the caller join this again?

Done.